File: solutions/clojure/bob/8/src/bob.clj
(ns bob
  (:require
   [clojure.string :as cstr]))

;; question -> Sure.
;; yell -> Whoa, chill out!
;; without saying -> Fine. Be that way!
;; else -> Whatever.

(def is-question? #(cstr/ends-with? % "?"))

(def is-forceful? 
  #(re-matches #"[A-Z]+"
              (cstr/join "" (re-seq #"[a-z|A-Z]+" %))))

(def all-spaces? #(empty? %))

(defn response-for
  [msg]
  (let [msg (cstr/trim msg)]
    (cond 
      (and
        (is-question? msg)
        (is-forceful? msg)) "Calm down, I know what I'm doing!"
      (is-forceful? msg)    "Whoa, chill out!"
      (is-question? msg)    "Sure."
      (all-spaces?  msg)    "Fine. Be that way!"
      :else "Whatever.")))

File: solutions/scala/bob/3/src/main/scala/Bob.scala
object Bob {
  def response(msg: String): String = {
    val trimmed = msg.trim

    val isSilent = (_: String).isEmpty
    def isShouting(s: String) = s.exists(_.isLetter) && s == s.toUpperCase
    def isQuestion(s: String) = s.endsWith("?")

    trimmed match {
      case s if isSilent(s) => "Fine. Be that way!"
      case s if isShouting(s) && isQuestion(s) => "Calm down, I know what I'm doing!"
      case s if isShouting(s) => "Whoa, chill out!"
      case s if isQuestion(s) => "Sure."
      case _ => "Whatever."
    }
  }
}
